pkg/systray/runtime/runner.go

```diff
@@ -0,0 +1,234 @@
+		level.Info(r.logger).Log(
+			"msg", "skipping systray for root user",
+			"uid", consoleOwnerUid,
+		)
+
+		return nil
+	}
+
+	uid := fmt.Sprintf("%d", consoleOwnerUid)
+
+	// already have a systray for the console owner
+	if _, ok := r.uidProcs[uid]; ok {
+		// if the process is still running, return
+		if isExists, _ := processExists(r.uidProcs[uid].Pid); isExists {
+			return nil
+		}
+
+		// proc is dead
+		level.Info(r.logger).Log(
+			"msg", "existing systray process dead for console user, starting new systray process",
+			"dead_pid", r.uidProcs[uid].Pid,
+			"uid", uid,
+		)
+	}
+
+	executable, err := os.Executable()
+	if err != nil {
+		return fmt.Errorf("getting executable path: %w", err)
+	}
+
+	proc, err := runAsUser(uid, executable, "systray")
+	if err != nil {
+		return fmt.Errorf("running systray: %w", err)
+	}
+
+	level.Info(r.logger).Log(
+		"msg", "systray started",
+		"uid", uid,
+		"pid", proc.Pid,
+	)
+
+	r.uidProcs[uid] = proc
+	return nil
+}
+
+func consoleOwnerUid() (uint32, error) {
+	const consolePath = "/dev/console"
+	consoleInfo, err := os.Stat(consolePath)
+	if err != nil {
+		return uint32(0), fmt.Errorf("stat %s: %w", consolePath, err)
+	}
+
+	return consoleInfo.Sys().(*syscall.Stat_t).Uid, nil
+}
+
+func runAsUser(uid string, path string, args ...string) (*os.Process, error) {
+	currentUser, err := user.Current()
+	if err != nil {
+		return nil, fmt.Errorf("getting current user: %w", err)
+	}
+
+	runningUser, err := user.LookupId(uid)
+	if err != nil {
+		return nil, fmt.Errorf("looking up user with uid %s: %w", uid, err)
+	}
+
+	cmd := exec.Command(path, args...)
+
+	// current user not root
+	if currentUser.Uid != "0" {
+
+		// if the user is running for itself, just run without setting credentials
+		if currentUser.Uid == uid {
+			err := cmd.Start()
+			if err != nil {
+				return nil, fmt.Errorf("running command: %w", err)
+			}
+			return cmd.Process, nil
+		}
+
+		// if the user is running for another user, we have an error because we can't set credentials
+		return nil, fmt.Errorf("current user %s is not root and cant start process for other user %s", currentUser.Uid, uid)
+	}
+
+	// the remaining code in this function is not covered by unit test since it requires root privileges
+	// We may be able to run passwordless sudo in GitHub actions, could possibly exec the tests as sudo.
+	// But we may not have a console user?
+
+	uidInt, err := strconv.Atoi(runningUser.Uid)
+	if err != nil {
+		return nil, fmt.Errorf("converting uid to int: %w", err)
+	}
+
+	gid, err := strconv.Atoi(runningUser.Gid)
+	if err != nil {
+		return nil, fmt.Errorf("converting gid to int: %w", err)
+	}
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Credential: &syscall.Credential{
+			Uid: uint32(uidInt),
+			Gid: uint32(gid),
+		},
+	}
+	err = cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("starting command: %w", err)
+	}
+
+	return cmd.Process, nil
+}
+
+func processExists(pid int) (bool, error) {
```

Comment:
Is it meaningful to return the error?

Reply:
yea, the error is just confusing, dropped it